### PR TITLE
SONAR-7054 use ws-client in batch

### PR DESCRIPTION
**Pull request by simonbrandhof**
SONAR-7054 use ws-client in batch

### CODE REVIEW

File: it/it-tests/src/test/java/it/authorisation/AuthenticationTest.java [hunk 1]
```diff
@@ -89,9 +91,9 @@ public void basic_authentication_based_on_login_and_password() {
     createUser(login, name, password);
 
     // authenticate
-    WsClient wsClient = new WsClient(newHttpConnector().url(ORCHESTRATOR.getServer().getUrl()).login(login).password(password).build());
```
Comment (teryk) [inline it/it-tests/src/test/java/it/authorisation/AuthenticationTest.java]: - I'd prefer a static method than `HttpConnector.Builder()`, maybe newHttpConnectorBuilder()
- +1 for replacing `login(string)` and `passowrd(string)` by `credentials(string, string)`
- as discussed, I'd set the url directly in the http connector constructor directly, as a mandatory field. Out of scope of this PR

File: it/it-tests/src/test/java/it/authorisation/AuthenticationTest.java [hunk 2]
```diff
@@ -89,9 +91,9 @@ public void basic_authentication_based_on_login_and_password() {
...
-    String response = wsClient.execute(newGetRequest("api/authentication/validate"));
```
Comment (teryk) [inline it/it-tests/src/test/java/it/authorisation/AuthenticationTest.java]: I prefer to hide the WsConnector when using the WsClient. It's used for initialization, but after that I'd much prefer a wsClient.call(WsRequest).
The same, I prefer static constructor method than new GetRequest and new PostRequest objects.

File: it/it-tests/src/test/java/it/authorisation/AuthenticationTest.java [hunk 3]
```diff
@@ -99,14 +101,13 @@ public void basic_authentication_based_on_token() {
     WsUserTokens.GenerateWsResponse generateWsResponse = userTokensWsClient.generate(new GenerateWsRequest()
       .setLogin(LOGIN)
       .setName("Validate token based authentication"));
-    WsClient wsClient = new WsClient(newHttpConnector()
+    WsClient wsClient = new HttpWsClient(new HttpConnector.Builder()
       .url(ORCHESTRATOR.getServer().getUrl())
-      .login(generateWsResponse.getToken())
-      .password("").build());
+      .token(generateWsResponse.getToken()).build());
```
Comment (teryk) [inline it/it-tests/src/test/java/it/authorisation/AuthenticationTest.java]: +1

File: pom.xml [hunk 4]
```diff
@@ -1367,7 +1365,7 @@
             <artifactId>maven-antrun-plugin</artifactId>
             <executions>
               <execution>
-                <id>protobuf-compile</id>
+                <id>compile-protobuf-sources</id>
```
Comment (teryk) [inline pom.xml]: for my knowledge, what does it change?

File: server/sonar-server/src/main/java/org/sonar/server/batch/GlobalAction.java [hunk 5]
```diff
@@ -63,7 +63,7 @@ public void handle(Request request, Response response) throws Exception {
     boolean hasScanPerm = userSession.hasGlobalPermission(GlobalPermissions.SCAN_EXECUTION);
     boolean hasPreviewPerm = userSession.hasGlobalPermission(GlobalPermissions.PREVIEW_EXECUTION);
     if (!hasPreviewPerm && !hasScanPerm) {
-      throw new ForbiddenException(Messages.NO_PERMISSION);
+      throw new ForbiddenException("You're not authorized to execute any SonarQube analysis. Please contact your SonarQube administrator.");
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/batch/GlobalAction.java]: why remove the constant?

File: sonar-batch/src/main/java/org/sonar/batch/cache/WSLoader.java [hunk 6]
```diff
@@ -101,7 +99,7 @@ public InputStream load(String id) throws IOException {
     }
   };
 
-  private class NotAvailableException extends Exception {
+  private static class NotAvailableException extends Exception {
```
Comment (teryk) [inline sonar-batch/src/main/java/org/sonar/batch/cache/WSLoader.java]: Shouldn't it extend `RuntimeException` ?

File: sonar-batch/src/main/java/org/sonar/batch/report/ReportPublisher.java [hunk 7]
```diff
@@ -141,129 +133,81 @@ private File prepareReport() {
...
-      String taskUrl = null;
-      if (taskId != null) {
-        taskUrl = baseUrl + "api/ce/task?id=" + taskId;
-        logger.info("More about the report processing at {}", taskUrl);
-      }
-
-      writeJson(url, taskUrl);
+      metadata.put("projectKey", effectiveKey);
+
+      URL dashboardUrl = HttpUrl.parse(publicUrl()).newBuilder()
```
Comment (teryk) [inline sonar-batch/src/main/java/org/sonar/batch/report/ReportPublisher.java]: +1

File: sonar-batch/src/main/java/org/sonar/batch/repository/DefaultQualityProfileLoader.java [hunk 8]
```diff
@@ -62,7 +62,7 @@ public DefaultQualityProfileLoader(WSLoader wsLoader) {
     return loadResource(url, fromCache);
   }
 
-  private static void validate(Collection<QualityProfile> profiles) {
+  private static void validate(@Nullable Collection<QualityProfile> profiles) {
     if (profiles == null || profiles.isEmpty()) {
```
Comment (teryk) [inline sonar-batch/src/main/java/org/sonar/batch/repository/DefaultQualityProfileLoader.java]: @henryju could be replaced by a `checkState` from Guava

File: sonar-batch/src/test/java/org/sonar/batch/bootstrap/WsClientLoggingInterceptorTest.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    List<String> debugLogs = logTester.logs(LoggerLevel.DEBUG);
+    assertThat(debugLogs).hasSize(1);
+    assertThat(debugLogs.get(0)).contains("GET 200 https://localhost:9000/api/issues/search | time=");
+    List<String> traceLogs = logTester.logs(LoggerLevel.TRACE);
+    assertThat(traceLogs).hasSize(1);
+    assertThat(traceLogs.get(0)).isEqualTo("GET https://localhost:9000/api/issues/search");
+  }
+
+  @Test
+  public void fail_if_requires_authentication() throws Exception {
```
Comment (teryk) [inline sonar-batch/src/test/java/org/sonar/batch/bootstrap/WsClientLoggingInterceptorTest.java]: why not use `ExpectedException` ?

File: sonar-batch/src/test/java/org/sonar/batch/bootstrap/WsClientLoggingInterceptorTest.java [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+    Request request = new Request.Builder()
+      .url("https://localhost:9000/api/issues/search")
+      .header("Authorization", "Basic BAD_CREDENTIALS")
+      .get()
+      .build();
+    Response response = newResponse(request, 401, "");
+    when(chain.request()).thenReturn(request);
+    when(chain.proceed(request)).thenReturn(response);
+
+    try {
```
Comment (teryk) [inline sonar-batch/src/test/java/org/sonar/batch/bootstrap/WsClientLoggingInterceptorTest.java]: `ExceptedException` ?

File: sonar-batch/src/test/java/org/sonar/batch/bootstrap/WsClientLoggingInterceptorTest.java [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+  }
+
+  @Test
+  public void fail_if_requires_permission() throws Exception {
+    Request request = newRequest();
+    Response response = newResponse(request, 403, "{\"errors\":[{\"msg\":\"missing scan permission\"}, {\"msg\":\"missing another permission\"}]}");
+    when(chain.request()).thenReturn(request);
+    when(chain.proceed(request)).thenReturn(response);
+
+    try {
```
Comment (teryk) [inline sonar-batch/src/test/java/org/sonar/batch/bootstrap/WsClientLoggingInterceptorTest.java]: ExceptedException ?

File: sonar-batch/src/test/java/org/sonar/batch/bootstrap/WsClientProviderTest.java [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+  @Test
+  public void provide_client_with_default_settings() {
+    GlobalProperties settings = new GlobalProperties(new HashMap<String, String>());
+
+    WsClient client = underTest.provide(settings, env);
+
+    assertThat(client).isNotNull();
+    HttpConnector httpConnector = (HttpConnector) client.wsConnector();
+    assertThat(httpConnector.baseUrl()).isEqualTo("http://localhost:9000/");
+    assertThat(httpConnector.okHttpClient().getProxy()).isNull();
```
Comment (teryk) [inline sonar-batch/src/test/java/org/sonar/batch/bootstrap/WsClientProviderTest.java]: do you really want to expose the ok http client publicly?

File: sonar-ws/src/main/java/org/sonarqube/ws/client/BaseRequest.java [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+import static java.util.Objects.requireNonNull;
+
+abstract class BaseRequest<SELF extends BaseRequest> implements WsRequest {
+
+  private final String path;
+
+  private String mediaType = MediaTypes.JSON;
+
+  // keep the same order -> do not use HashMap
+  private final Map<String, String> params = new LinkedHashMap<>();
```
Comment (teryk) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/BaseRequest.java]: +1 (shame on me)

File: sonar-ws/src/main/java/org/sonarqube/ws/client/GetRequest.java [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarqube.ws.client;
+
+/**
+ * @since 5.3
+ */
+public class GetRequest extends BaseRequest<GetRequest> {
```
Comment (teryk) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/GetRequest.java]: Now I understand your point with the GetRequest etc. It's more elegant, but I would still hide it from the user of WsClient. I'd keep the newGetRequest and newPostRequest from the WsRequest class. WDYT?

File: sonar-ws/src/main/java/org/sonarqube/ws/client/HttpConnector.java [hunk 15]
```diff
@@ -17,73 +17,192 @@
...
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static com.google.common.base.Strings.nullToEmpty;
+import static java.lang.String.format;
+
 public class HttpConnector implements WsConnector {
 
-  public static final int DEFAULT_CONNECT_TIMEOUT_MILLISECONDS = 30000;
-  public static final int DEFAULT_READ_TIMEOUT_MILLISECONDS = 60000;
+  public static final int DEFAULT_CONNECT_TIMEOUT_MILLISECONDS = 30_000;
```
Comment (teryk) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/HttpConnector.java]: +1

File: sonar-ws/src/main/java/org/sonarqube/ws/client/HttpConnector.java [hunk 16]
```diff
@@ -17,73 +17,192 @@
...
   }
 
   @Override
-  public String execute(WsRequest wsRequest) {
-    return requestFactory.execute(wsRequest);
+  public String baseUrl() {
+    return baseUrl.url().toExternalForm();
+  }
+
+  public OkHttpClient okHttpClient() {
```
Comment (teryk) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/HttpConnector.java]: as said previously, I wouldn't expose the internal http client, specific to a library

File: sonar-batch/src/main/java/org/sonar/batch/bootstrap/WsClientProvider.java [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+  static final int DEFAULT_READ_TIMEOUT_SEC = 60;
+
+  private HttpWsClient wsClient;
+
+  public synchronized WsClient provide(final GlobalProperties settings, final EnvironmentInformation env) {
+    if (wsClient == null) {
+      String url = defaultIfBlank(settings.property("sonar.host.url"), CoreProperties.SERVER_BASE_URL_DEFAULT_VALUE);
+      HttpConnector.Builder builder = new HttpConnector.Builder();
+
+      // TODO proxy
```
Comment (SonarTech) [inline sonar-batch/src/main/java/org/sonar/batch/bootstrap/WsClientProvider.java]: INFO Complete the task associated to this TODO comment. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1135)

File: sonar-batch/src/main/java/org/sonar/batch/bootstrap/WsClientLoggingInterceptor.java [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+    try {
+      JsonParser parser = new JsonParser();
+      JsonObject obj = parser.parse(responseContent).getAsJsonObject();
+      JsonArray errors = obj.getAsJsonArray("errors");
+      List<String> errorMessages = new ArrayList<>();
+      for (JsonElement e : errors) {
+        errorMessages.add(e.getAsJsonObject().get("msg").getAsString());
+      }
+      return Joiner.on(", ").join(errorMessages);
+    } catch (Exception e) {
```
Comment (SonarTech) [inline sonar-batch/src/main/java/org/sonar/batch/bootstrap/WsClientLoggingInterceptor.java]: CRITICAL Either log or rethrow this exception. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1166)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* CRITICAL 1 critical
* MAJOR 2 major
* INFO 1 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Protected method 'doBeforeStart(...)' is never used.](https://github.com/SonarSource/sonarqube/blob/d6b0df77736545d7dcfb5400fe937fd6a309a278/sonar-batch/src/main/java/org/sonar/batch/bootstrap/GlobalContainer.java#L63) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUnusedProtectedMethod)
* MAJOR [Protected method 'doAfterStart(...)' is never used.](https://github.com/SonarSource/sonarqube/blob/d6b0df77736545d7dcfb5400fe937fd6a309a278/sonar-batch/src/main/java/org/sonar/batch/bootstrap/GlobalContainer.java#L109) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUnusedProtectedMethod)

File: sonar-ws/src/main/java/org/sonarqube/ws/client/HttpResponse.java [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+  }
+
+  @Override
+  public boolean hasContent() {
+    return okResponse.code() != HTTP_NO_CONTENT;
+  }
+
+  @Override
+  public String getContentType() {
+    return okResponse.header("Content-Type");
```
Comment (teryk) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/HttpResponse.java]: Please use https://hc.apache.org/httpcomponents-core-ga/httpcore/apidocs/constant-values.html#org.apache.http.HttpHeaders.CONTENT_TYPE or another alternative

File: sonar-ws/src/main/java/org/sonarqube/ws/client/PostRequest.java [hunk 20]
```diff
@@ -17,55 +17,53 @@
...
   @Override
-  protected void after() {
-    server.stop();
+  public Method getMethod() {
+    return Method.POST;
   }
 
-  public MockHttpServerInterceptor stubResponseBody(String body) {
-    server.doReturnBody(body);
+  public PostRequest setPart(String name, Part part) {
```
Comment (teryk) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/PostRequest.java]: rename to addPart?

File: sonar-ws/src/main/java/org/sonarqube/ws/client/WsClient.java [hunk 21]
```diff
@@ -17,74 +17,30 @@
...
 
-  public IssuesWsClient issuesWsClient() {
-    return issuesWsClient;
-  }
+  UserTokensService userTokens();
 
-  public UserTokensWsClient userTokensWsClient() {
-    return userTokensWsClient;
-  }
+  WsConnector wsConnector();
```
Comment (teryk) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/WsClient.java]: +1 HttpConnector -> WsConnector

**Comment by teryk** [comment]
not in the context of this PR, but the same must be done on 2 plugins to handle the empty password case

File: server/sonar-server/src/main/java/org/sonar/server/batch/GlobalAction.java (hunk 5 above)
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/batch/GlobalAction.java]: I don't know. My bad.

File: pom.xml (hunk 4 above)
Comment (simonbrandhof) [inline pom.xml]: all other execution ids in the pom have the form "do something"

File: sonar-batch/src/main/java/org/sonar/batch/cache/WSLoader.java (hunk 6 above)
Comment (simonbrandhof) [inline sonar-batch/src/main/java/org/sonar/batch/cache/WSLoader.java]: It's ouf of scope of this PR, but I think this is used for control flow...

File: sonar-batch/src/test/java/org/sonar/batch/bootstrap/WsClientLoggingInterceptorTest.java (hunk 9 above)
Comment (simonbrandhof) [inline sonar-batch/src/test/java/org/sonar/batch/bootstrap/WsClientLoggingInterceptorTest.java]: Fixed

File: sonar-batch/src/test/java/org/sonar/batch/bootstrap/WsClientProviderTest.java (hunk 12 above)
Comment (simonbrandhof) [inline sonar-batch/src/test/java/org/sonar/batch/bootstrap/WsClientProviderTest.java]: I'm ok as this is in the implementation but not in the interface WsConnector. In any case okhttp is a transitive dep of ws-client.

File: sonar-ws/src/main/java/org/sonarqube/ws/client/HttpResponse.java (hunk 19 above)
Comment (simonbrandhof) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/HttpResponse.java]: Of course we won't add the dependency to Apache HttpClient just for this constant. I don't know any other alternatives. Okhttp does not have such constants.

File: sonar-ws/src/main/java/org/sonarqube/ws/client/PostRequest.java (hunk 20 above)
Comment (simonbrandhof) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/PostRequest.java]: why ? It replaces existing key and method name for string params is already `setParam()`.

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* CRITICAL 1 critical
* MAJOR 2 major
* INFO 1 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Protected method 'doBeforeStart(...)' is never used.](https://github.com/SonarSource/sonarqube/blob/bd54c2aa8754cead7d4b6810a08c2bebd2488c77/sonar-batch/src/main/java/org/sonar/batch/bootstrap/GlobalContainer.java#L63) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUnusedProtectedMethod)
* MAJOR [Protected method 'doAfterStart(...)' is never used.](https://github.com/SonarSource/sonarqube/blob/bd54c2aa8754cead7d4b6810a08c2bebd2488c77/sonar-batch/src/main/java/org/sonar/batch/bootstrap/GlobalContainer.java#L109) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUnusedProtectedMethod)

**Pull request closed without merging**